internal/vm/task.go

```diff
@@ -0,0 +1,338 @@
+	errCh := make(chan error, 3)
+
+	// For each of stdin, stdout and stderr, if a FIFO has been set (i.e. the client requested that stream to be setup), go
+	// open the FIFO and connect the vsock.
+	if task.fifoSet.Stdin != "" {
+		wg.Add(1)
+		go func() {
+			errCh <- task.proxyIO(ctx, task.fifoSet.Stdin, task.extraData.StdinPort, inputDirection, vsockConnector)
+			wg.Done()
+		}()
+	} else {
+		logger.Info("skipping proxy io for unset stdin")
+	}
+
+	if task.fifoSet.Stdout != "" {
+		wg.Add(1)
+		go func() {
+			errCh <- task.proxyIO(ctx, task.fifoSet.Stdout, task.extraData.StdoutPort, inputDirection.opposite(), vsockConnector)
+			wg.Done()
+		}()
+	} else {
+		logger.Info("skipping proxy io for unset stdout")
+	}
+
+	if task.fifoSet.Stderr != "" {
+		wg.Add(1)
+		go func() {
+			errCh <- task.proxyIO(ctx, task.fifoSet.Stderr, task.extraData.StderrPort, inputDirection.opposite(), vsockConnector)
+			wg.Done()
+		}()
+	} else {
+		logger.Info("skipping proxy io for unset stderr")
+	}
+
+	// Once the streams have been initialized, gather any errors they returned and let the caller know by publishing
+	// on the return chan
+	go func() {
+		wg.Wait()
+		close(errCh)
+
+		var finalError *multierror.Error
+		for err := range errCh {
+			finalError = multierror.Append(finalError, err)
+		}
+
+		returnCh <- finalError.ErrorOrNil()
+		close(returnCh)
+	}()
+
+	return returnCh
+}
+
+func (task *Task) proxyIO(ctx context.Context, fifoPath string, port uint32, ioDirection IODirection, vsockConnector VSockConnector) error {
+	logger := log.G(ctx).WithField("path", fifoPath).WithField("port", port)
+
+	fifoFile, err := fifo.OpenFifo(ctx, fifoPath, syscall.O_RDWR|syscall.O_CREAT|syscall.O_NONBLOCK, 0700)
+	if err != nil {
+		logger.WithError(err).Error("failed to open fifo")
+		return err
+	}
+
+	sock, err := vsockConnector(ctx, port)
+	if err != nil {
+		fifoFile.Close()
+		logger.WithError(err).Error("failed to establish vsock connection")
+		return err
+	}
+
+	// the FIFO and vsock have been opened, so go off proxying between them asynchronously
+	// until Cancel is called.
+	go func() {
+		buf := make([]byte, internal.DefaultBufferSize)
+
+		logger.Debug("begin copying io")
+
+		switch ioDirection {
+		case FIFOtoVSock:
+			go func() {
+				<-task.cancelCh
+				// If we are copying from FIFO to vsock, be sure to first close just the FIFO,
+				// allowing any buffered data to be flushed into the vsock first.
+				fifoFile.Close()
+			}()
+
+			_, err = io.CopyBuffer(sock, fifoFile, buf)
+			sock.Close()
+
+		case VSockToFIFO:
+			go func() {
+				<-task.cancelCh
+				// If we are copying from vsock to FIFO, be sure to first close just the vsock,
+				// allowing any buffered data to be flushed into the FIFO first.
+				sock.Close()
+			}()
+
+			_, err = io.CopyBuffer(fifoFile, sock, buf)
+			fifoFile.Close()
+
+		default:
+			logger.Fatalf("undefined io direction") // should not be possible, direction is a bool
+		}
+
+		if err != nil {
```

Comment:
`fifoFile` is closed in all the previous code branches already before this line

Reply:
Yea, I figured as much but would it make sense to remove the branch closes for a single close in this if statement?

Reply:
In this situation, no I don't think so. We want to close the FIFO in a very particular order that differs between the two case statements above this.

When we are copying `FIFOtoVsock`, we want to close the FIFO before the vsock. This allows us to first cut off the FIFO, then drain buffered data and then finally close the vsock.

When we are copying `VSockToFIFO`, we the exact opposite, to first cut off the vsock, then drain buffered data and then finally close the FIFO.

Reply:
Ah, I see. Hm, Im just worried about adding another switch case and having someone forget to close the fifo. Can we add a comment illustrating that each case in the switch statement should close the fifo since that isn't handled outside of the switch scope.

Reply:
There shouldn't be any possibility of adding another case as the underlying type here is a `bool`, so there can only be 2 cases.

Adding the comment though for extra clarity.